### PR TITLE
added ability to opt out from forward/backwards cursor pagination

### PR DESCRIPTION
**Pull request by icanhasjonas**
Added ability to opt-out to from backward or forward pagination

```cs
// Forward and Backward defaults to `true`
[UsePaging( Backward = false, Foward = true )]
... 
```

### CODE REVIEW

**Comment by michaelstaib** [comment]
Hi,

thanks for taking the time to put together a pull-request.

I got a couple of questions:
- What is the use-case you see behind this?
- What happens when forward and backward are set to true?
- What happens with forward and backward when options are used in an offset context?

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs [hunk 1]
```diff
@@ -85,33 +85,63 @@ public static class PagingObjectFieldDescriptorExtensions
...
-                .Argument(CursorPagingArgumentNames.First, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.After, a => a.Type<StringType>())
-                .Argument(CursorPagingArgumentNames.Last, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.Before, a => a.Type<StringType>());
+            if (forward)
+            {
+                descriptor = descriptor
+                        .Argument(CursorPagingArgumentNames.First, a => a.Type<IntType>())
+                        .Argument(CursorPagingArgumentNames.After, a => a.Type<StringType>())
+                    ;
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs:102]: The code must be reformatted.

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs [hunk 2]
```diff
@@ -85,33 +85,63 @@ public static class PagingObjectFieldDescriptorExtensions
...
+            }
+
+            if (backward)
+            {
+                descriptor = descriptor
+                        .Argument(CursorPagingArgumentNames.Last, a => a.Type<IntType>())
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs:108]: The code must be reformatted.

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs [hunk 3]
```diff
@@ -85,33 +85,63 @@ public static class PagingObjectFieldDescriptorExtensions
...
 
-            return descriptor
-                .Argument(CursorPagingArgumentNames.First, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.After, a => a.Type<StringType>())
-                .Argument(CursorPagingArgumentNames.Last, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.Before, a => a.Type<StringType>());
+            if (forward)
+            {
+                descriptor = descriptor
+                        .Argument(CursorPagingArgumentNames.First, a => a.Type<IntType>())
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs:130]: The code must be reformatted.

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs [hunk 4]
```diff
@@ -85,33 +85,63 @@ public static class PagingObjectFieldDescriptorExtensions
...
+                        .Argument(CursorPagingArgumentNames.After, a => a.Type<StringType>())
+                    ;
+            }
+
+            if (backward)
+            {
+                descriptor = descriptor
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs:137]: The code must be reformatted.

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs [hunk 5]
```diff
@@ -85,33 +85,63 @@ public static class PagingObjectFieldDescriptorExtensions
...
             {
                 throw new ArgumentNullException(nameof(descriptor));
             }
 
-            return descriptor
-                .Argument(CursorPagingArgumentNames.First, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.After, a => a.Type<StringType>())
-                .Argument(CursorPagingArgumentNames.Last, a => a.Type<IntType>())
-                .Argument(CursorPagingArgumentNames.Before, a => a.Type<StringType>());
+            if (forward)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/PagingObjectFieldDescriptorExtensions.cs:127]: What if both forward and backward are false?

File: src/HotChocolate/Core/src/Types.CursorPagination/Extensions/UsePagingAttribute.cs [hunk 6]
```diff
@@ -63,6 +65,24 @@ public bool IncludeTotalCount
             set => _includeTotalCount = value;
         }
 
+        /// <summary>
+        /// Allow forward paging using <c>first</c> and <c>after</c>
+        /// </summary>
+        public bool Forward
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.CursorPagination/Extensions/UsePagingAttribute.cs:71]: Shouldn^t this be called AllowForwardNavigation? Or something in this direction?

File: src/HotChocolate/Core/src/Types/Types/Pagination/PagingOptions.cs [hunk 7]
```diff
@@ -7,5 +7,9 @@ public struct PagingOptions
         public int? MaxPageSize { get; set; }
 
         public bool? IncludeTotalCount { get; set; }
+
+        public bool Forward { get; set; }
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/Pagination/PagingOptions.cs:11]: The paging options are meant for both offset and cursor pagination ... if we have some more specialized cases we might want to have an inherited version.

@PascalSenn this again bears the question of how we proceed with pagination.

**Comment by icanhasjonas** [comment]
> Hi,
> 
> thanks for taking the time to put together a pull-request.
> 
> I got a couple of questions:
> 
> * What is the use-case you see behind this?

Some data sources don't behave well on backward paging, and counting, such as DynamoDB and ElasticSearch, S3 and others

> * What happens when forward and backward are set to true?

That's the normal case, no changes.

> * What happens with forward and backward when options are used in an offset context?

This is one I didn't anticipate, tbh. I suppose in its current state, `forward` and `backward` are not compatible with the offset paging scenario

**Comment by icanhasjonas** [comment]
After some thoughts, maybe a better solution is to have the implemented `IPagingProvider` call `descriptor.AddPagingArguments()` with suggested capabilities (such as forward and backward abilities, translating into `first`, `last`, `after`, `before`). That way I can simply register a ForwardOnly or DynamoDb `CursorPagingHandler`.

**Comment by michaelstaib** [comment]
So actually we want to have backward pagination optional?

Since the paging arguments are specified by the relay paginations spec I think allowing the provider could lead to unwanted side-effects.

**Comment by michaelstaib** [comment]
Why I am a bit cautious about adding both is that I try to tend to only add things that have a real use-case. I can see that for disabling backward navigation.

In your case you really want a forward only navigation .... I can see that also as something valid for things like Microsoft table storage and so on. 

Do you agree?

**Comment by michaelstaib** [comment]
Also on the plus side, if we only allow for disabling backward navigation we will no longer run into errors when both is disabled. Meaning we do not need to add any error handling on this.

**Comment by icanhasjonas** [comment]
Agreed, that's a good point.

**Comment by michaelstaib** [comment]
We will still have a look at this ... we have around 2 weeks left off client work after that we will have a look at this.

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**